pkg/vault/contracts/AggregatorBatchRouter.sol

```diff
@@ -0,0 +1,476 @@
+                    abi.encodeCall(
+                        AggregatorBatchRouter.swapExactOutHook,
+                        SwapExactOutHookParams({
+                            sender: msg.sender,
+                            paths: paths,
+                            deadline: deadline,
+                            wethIsEth: wethIsEth,
+                            userData: userData
+                        })
+                    )
+                ),
+                (uint256[], address[], uint256[])
+            );
+    }
+
+    function swapExactInHook(
+        SwapExactInHookParams calldata params
+    )
+        external
+        nonReentrant
+        onlyVault
+        returns (uint256[] memory pathAmountsOut, address[] memory tokensOut, uint256[] memory amountsOut)
+    {
+        (pathAmountsOut, tokensOut, amountsOut) = _swapExactInHook(params);
+
+        _settlePaths(params.sender, params.wethIsEth);
+    }
+
+    function _swapExactInHook(
+        SwapExactInHookParams calldata params
+    ) internal returns (uint256[] memory pathAmountsOut, address[] memory tokensOut, uint256[] memory amountsOut) {
+        // The deadline is timestamp-based: it should not be relied upon for sub-minute accuracy.
+        // solhint-disable-next-line not-rely-on-time
+        if (block.timestamp > params.deadline) {
+            revert SwapDeadline();
+        }
+
+        pathAmountsOut = _computePathAmountsOut(params);
+
+        // The hook writes current swap token and token amounts out.
+        // We copy that information to memory to return it before it is deleted during settlement.
+        tokensOut = _currentSwapTokensOut().values();
+        amountsOut = new uint256[](tokensOut.length);
+        for (uint256 i = 0; i < tokensOut.length; ++i) {
+            amountsOut[i] = _currentSwapTokenOutAmounts().tGet(tokensOut[i]);
+        }
+    }
+
+    function _computePathAmountsOut(
+        SwapExactInHookParams calldata params
+    ) internal returns (uint256[] memory pathAmountsOut) {
+        pathAmountsOut = new uint256[](params.paths.length);
+
+        if (EVMCallModeHelpers.isStaticCall() == false) {
+            for (uint256 i = 0; i < params.paths.length; ++i) {
+                SwapPathExactAmountIn memory path = params.paths[i];
+                _currentSwapTokensIn().add(address(path.tokenIn));
+                _currentSwapTokenInAmounts().tAdd(address(path.tokenIn), path.exactAmountIn);
+            }
+
+            address[] memory tokensIn = _currentSwapTokensIn().values();
+            for (uint256 i = 0; i < tokensIn.length; ++i) {
+                uint256 amount = _currentSwapTokenInAmounts().tGet(tokensIn[i]);
+                uint256 tokenInCredit = _vault.settle(IERC20(tokensIn[i]), amount);
+                if (tokenInCredit < amount) {
+                    revert InsufficientPayment();
+                }
+
+                _currentSwapTokenInAmounts().tSet(tokensIn[i], 0);
+            }
```

Comment:
This logic looks fine.

An alternative would be to just call settle for the token in for each of the paths though. The current approach is better if you have many paths using the same token in, as you minimize the calls to `vault.settle`. But if you don't have that situation, it might be better to just call `vault.settle` for each token in and call it a day. You wouldn't need to update token in amounts, tokens in, etc.

WDYT?

Reply:
Hmm, then we would need to limit the case when multiple same tokens are used, otherwise there will be an error if we try to settle the same token twice.